ceno_zkvm/src/scheme/verifier.rs

```diff
@@ -516,42 +518,49 @@ impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> ZKVMVerifier<E, PCS>
                 .zip_eq(cs.w_table_expressions.iter())
                 .all(|(r, w)| r.table_spec.len == w.table_spec.len)
         );
-        let is_skip_same_point_sumcheck = cs
-            .r_table_expressions
-            .iter()
-            .chain(cs.w_table_expressions.iter())
-            .map(|rw| rw.table_spec.len)
-            .chain(cs.lk_table_expressions.iter().map(|lk| lk.table_len))
-            .all_equal();
+        // in table proof, we always skip same point sumcheck for now
+        // as tower sumcheck batch product argument/logup in same length
+        let is_skip_same_point_sumcheck = true;
 
         // verify and reduce product tower sumcheck
         let tower_proofs = &proof.tower_proof;
 
-        let expected_rounds = izip!(
-            // w_table_expression round match with r_table_expression so it fine to check either of them
-            &cs.r_table_expressions,
-            &proof.rw_hints_num_vars
-        )
-        .flat_map(|(r, hint_num_vars)| match r.table_spec.addr_type {
-            // fixed address: get number of round from vk
-            SetTableAddrType::FixedAddr => {
-                let num_vars = ceil_log2(r.table_spec.len);
+        let expected_rounds = cs
+            .r_table_expressions
+            .iter()
+            .flat_map(|r| {
+                // iterate through structural witins and collect max round.
+                let Some(num_vars) = r.table_spec.len.map(ceil_log2).or_else(|| {
+                    r.table_spec
+                        .structural_witins
+                        .iter()
+                        .map(|structural_witin| {
+                            let Expression::StructuralWitIn(addr_witin_id, max_len, _, _) =
```

Comment:
a nitpick: better rename `addr_witin_id` just `witin_id` As here it can be more general concept